src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/XDataTypeTransformer.cs

```diff
@@ -0,0 +1,79 @@
+using System.Collections.Generic;
+using System.Linq;
+using XamlX;
+using XamlX.Ast;
+using XamlX.Transform;
+using XamlX.Transform.Transformers;
+using XamlX.TypeSystem;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers
+{
+    internal class XDataTypeTransformer : IXamlAstTransformer
+    {
+        private const string DataTypePropertyName = "DataType";
+        
+        /// <summary>
+        /// Converts x:DataType directives to regular DataType assignments if property with Avalonia.Metadata.DataTypeAttribute exists.
+        /// </summary>
+        /// <returns></returns>
+        public IXamlAstNode Transform(AstTransformationContext context, IXamlAstNode node)
+        {
+            if (node is XamlAstObjectNode on)
+            {
+                for (var c = 0; c < on.Children.Count; c++)
+                {
+                    var ch = on.Children[c];
+                    if (ch is XamlAstXmlDirective { Namespace: XamlNamespaces.Xaml2006, Name: DataTypePropertyName } d)
+                    {
+                        if (on.Children.OfType<XamlAstXamlPropertyValueNode>()
+                            .Any(p => ((XamlAstNamePropertyReference)p.Property)?.Name == DataTypePropertyName))
+                        {
+                            // Break iteration if any DataType property was already set by user code.
+                            break;
+                        }
+                        
+                        var templateDataTypeAttribute = context.GetAvaloniaTypes().DataTypeAttribute;
+
+                        var clrType = on.Type switch
+                        {
+                            XamlAstClrTypeReference clrRef => clrRef.Type,
+                            XamlAstXmlTypeReference xmlRef => TypeReferenceResolver.ResolveType(context, xmlRef.Name,
+                                on.Type.IsMarkupExtension, on, strict: false).Type,
+                            _ => null
+                        };
+                        if (clrType is null)
+                        {
+                            break;
+                        }
+
+                        // Technically it's possible to map "x:DataType" to a property with [DataType] attribute regardless of its name,
+                        // but we go explicitly strict here and check the name as well.
+                        var (declaringType, dataTypeProperty) = GetAllProperties(clrType)
```

Comment:
You are checking for existence of a property with `DataTypePropertyName` but later resolve property name via attribute and don't check if the property already exists.

Reply:
In both places [line 29](https://github.com/AvaloniaUI/Avalonia/pull/8220/files#diff-8aa4de7f0efe422d032b55106fcf255392c61f09c9a901b40417a0672f5e847eR29) and [line 46](https://github.com/AvaloniaUI/Avalonia/pull/8220/files#diff-8aa4de7f0efe422d032b55106fcf255392c61f09c9a901b40417a0672f5e847eR46) I check for DataTypePropertyName. Attribute is just additional condition that should be met.

Reply:
Added another test for that XDataType_Should_Be_Ignored_If_DataType_Has_Non_Standard_Name

Reply:
Is the property marked with `[DataType]` guaranteed to be named `DataType`?

Reply:
e. g. here you added `FancyDataType` - https://github.com/AvaloniaUI/Avalonia/blob/a1f8b5c9a7b9f016766c981a7a0ea2481e6c34c1/tests/Avalonia.Markup.Xaml.UnitTests/MarkupExtensions/CompiledBindingExtensionTests.cs#L1523-L1524

Reply:
Ah, nvm, it's a separate branch.